chapter/methodology.md

```diff
@@ -0,0 +1,964 @@
+
+To separate the phases of the elaboration of the linearization in a type-safe, the implementation is based on type-states[@typestate].
+Type-states were chosen over an enumeration bases approach for the additional flexibility they provide to build a generic interface.
+Thanks to the generic interface, the adaptions to Nickel to integrate NLS are expected to have almost no influence on the runtime performance of the language in an optimized build.
+
+NLS implements separate type-states for the two phases of the linearization: `Building` and `Completed`.
+
+
+building phase:
+  ~ A linearization in the `Building` state is a linearization under construction.
+    It is a list of `LinearizationItem`s of unresolved type, appended as they are created during a depth-first traversal of the AST.
+  ~ During this phase, the `id` affected to a new item is always equal to its index in the array.
+  ~ The Building state also records the definitions in scope of each item in a separate mapping.
+
+post-processing phase:
+  ~ Once fully built, a Building instance is post-processed to get a `Completed` linearization.
+  ~ Although fundamentally still represented by an array, a completed linearization is optimized for search by positions (in the source file) thanks to sorting and the use of an auxiliary map from `id`s to the new index of items.
+  ~ Additionally, missing edges in the usage graph have been created and he types of items are fully resolved in a completed linearization.
+
+Type definitions of the `Linearization` as well as its type-states `Building` and `Completed` are listed in [@lst:nickel-definition-lineatization;@lst:nls-definition-building-type;@lst:nls-definition-completed-type].
+Note that only the former is defined as part of the Nickel libraries, the latter are specific implementations for NLS.
+
+
+```{.rust #lst:nickel-definition-lineatization caption="Definition of Linearization structure"}
+pub trait LinearizationState {}
+
+pub struct Linearization<S: LinearizationState> {
+    pub state: S,
+}
+```
+
+```{.rust #lst:nls-definition-building-type caption="Type Definition of Building state"}
+pub struct Building {
+    pub linearization: Vec<LinearizationItem<Unresolved>>,
+    pub scope: HashMap<Vec<ScopeId>, Vec<ID>>,
+}
+impl LinearizationState for Building {}
+```
+
+```{.rust #lst:nls-definition-completed-type caption="Type Definition of Completed state" }
+pub struct Completed {
+    pub linearization: Vec<LinearizationItem<Resolved>>,
+    scope: HashMap<Vec<ScopeId>, Vec<ID>>,
+    id_to_index: HashMap<ID, usize>,
+}
+impl LinearizationState for Completed {}
+```
+
+### Transfer from AST
+
+The NLS project aims to present a transferable architecture that can be adapted for future languages.
+Consequently, NLS faces the challenge of satisfying multiple goals
+
+1. To keep up with the frequent changes to the Nickel language and ensure compatibility at minimal cost, NLS needs to *integrate critical functions* of Nickel's runtime
```

Comment:
I don't know if this is the right place, but you can make the point that this worked well, because we butchered the syntax a lot those last 3 weeks (unified syntax, list -> array, changing comments, changing interpolation, enum types, etc.) and the LSP just kept working, basically.

Reply:
Maybe better to put this into #8